src/hotspot/os/linux/os_linux.cpp

```diff
@@ -216,18 +216,18 @@ julong os::Linux::available_memory() {
 }
 
 julong os::physical_memory() {
-  jlong phys_mem = 0;
+  jlong phys_mem = Linux::physical_memory();
+
   if (OSContainer::is_containerized()) {
     jlong mem_limit;
-    if ((mem_limit = OSContainer::memory_limit_in_bytes()) > 0) {
+    if ((mem_limit = OSContainer::memory_limit_in_bytes()) > 0 && mem_limit < phys_mem) {
```

Comment:
I don't understand this, but also not the preexisting code.

Did we not just *overwrite* `os::Linux::_physical_memory` during the initialization of the container subsystem depending on the limit? Why is there even a need for added logic here, why not just return `os::Linux::_physical_memory`?

Maybe the answer is that `os::physical_memory()` is supposed to mirror cgroup limit changes during the VMs lifetime (is it?). But then, the question is the other way around, why bother correcting the value during initialization if we need to correct it in `os::physical_memory()` ?

Reply:
Good questions. I think the answer is that the cgroups values might change during VMs lifetime. I never understood why that was a goal to support, but fine. It's possible to do via docker so I guess that's why. As to the question why bother doing it in init too? I don't know. Maybe @dholmes-ora remembers?

Perhaps the better fix would be to fix the values in `OSContainer::memory_limit_in_bytes()`, instead of at all call-sites. It appears `os::Linux::available_memory()` would need similar treatment.

Reply:
@jerboaa I don't  sorry. My involvement with the container support was only peripheral.

Reply:
Looking at the [original changeset](https://hg.openjdk.java.net/jdk/jdk/rev/7f22774a5f42#l11.30) for this it appears the only reason for it is to account for `-XX:+AggressiveHeap` and setting related flags. As to why the hook in `os::physical_memory()` isn't sufficient I'm at a loss. The original review here https://mail.openjdk.org/pipermail/hotspot-dev/2017-October/028635.html suggests that there was supposed to be some support for interleaving os initialization and arguments processing, but I don't see it. To be safe, I'd just leave this as it is for this patch and potentially clean this up as a separate bug. Thoughts?

Reply:
Okay. Maybe something for a follow-up RFE.

Reply:
I've tried to address this in 4d8a85732c2ebbe31ae1dfc5c756ac14d4989755. My approach was to bubble the physical-ram-versus-cgroup-limit comparison right up to `os::physical_memory`. Now, the cgroups code simply reads and reports what is in cgroups (even if its nonsense) and we no longer overwrite `Linux::_physical_memory` with the cgroup value (which felt misleading to me, and also, meant the physical host RAM value is lost to the JVM after the first cgroup read: if the JVM launched under a low cgroup memory limit that was later lifted, that higher number would never be reported)

Unfortunately, the logging is quite a lot noisier:

```
# baseline
$ /usr/lib/jvm/java-17-openjdk-amd64/bin/java -Xlog:os+container=trace -XX:InitialRAMPercentage=25.0 -XX:+PrintFlagsFinal -version 2>&1 | wc -l
724
$ /usr/lib/jvm/java-17-openjdk-amd64/bin/java -Xlog:os+container=trace -XX:InitialRAMPercentage=25.0 -XX:+PrintFlagsFinal -version 2>&1 | awk '{ $1=""; print }' | sort | uniq -c | sort -nk1 | tail
      3  Raw value for memory limit is: 137438953472
      4  Memory Usage is: 3886501888
      4  Open of file /sys/fs/cgroup/user.slice/user-1000.slice/user@1000.service/app.slice/app-org.gnome.Terminal.slice/vte-spawn-cdfa23a8-08e3-4781-b456-573c833609e9.scope/cpu.max failed, No such file or directory
      4  Path to /cpu.max is /sys/fs/cgroup/user.slice/user-1000.slice/user@1000.service/app.slice/app-org.gnome.Terminal.slice/vte-spawn-cdfa23a8-08e3-4781-b456-573c833609e9.scope/cpu.max
      6  Memory Usage is: 3887812608
      7  Memory Usage is: 3887157248
      8  Memory Usage is: 3888467968
      9  Memory Usage is: 3886239744
      9  Memory Usage is: 3888599040
     58  Path to /memory.current is /sys/fs/cgroup/user.slice/user-1000.slice/user@1000.service/app.slice/app-org.gnome.Terminal.slice/vte-spawn-cdfa23a8-08e3-4781-b456-573c833609e9.scope/memory.current
# this PR
$ $JAVA_HOME/bin/java -Xlog:os+container=trace -XX:InitialRAMPercentage=25.0 -XX:+PrintFlagsFinal -version 2>&1 | wc -l
1418
$ ;$JAVA_HOME/bin/java -Xlog:os+container=trace -XX:InitialRAMPercentage=25.0 -XX:+PrintFlagsFinal -version 2>&1 | awk '{ $1=""; print }' | sort | uniq -c | sort -nk1 | tail
     10  Memory Usage is: 3833311232
     11  Memory Usage is: 3826597888
     11  Memory Usage is: 3827486720
     12  Memory Usage is: 3828273152
     13  Memory Usage is: 3826860032
     14  Memory Usage is: 3827118080
     15  Memory Usage is: 3827748864
     17  Memory Usage is: 3825704960
     20  Memory Usage is: 3826339840
    404  Path to /memory.current is /sys/fs/cgroup/user.slice/user-1000.slice/user@1000.service/app.slice/app-org.gnome.Terminal.slice/vte-spawn-cdfa23a8-08e3-4781-b456-573c833609e9.scope/memory.current
```